### DISCUSSION TITLE
Escaping text for automatic insertion

### DISCUSSION THREAD

**Discussion by KillTheMule** [typst/typst#3773] [Q&A] [answered]
I'm templating documents with typst, and as such, I'm putting user-provided strings into typst files. That runs the risk of breaking the file, because the user might input a `#` or any special character recognized by typst.

So I probably need to escape the text I'm inserting. Is there anything to help me with it? I haven't found a function anywhere for that, so I guess I need to write my own. Maybe someone already did this? Is there a definitive list of what typst recognises as special, or should I just go by https://typst.app/docs/reference/syntax/#escapes?

Any ideas on how to do this most practically? Maybe I could just put the text in quotes and then escape the quote symbols? But then I'm back to trying out what exactly needs quoting here (there's a small list of escape sequences I need to deal with, but I'd be wary of just assuming that double slash doesn't start a comment even inside a quoted string).

Anyways, escaping text is always somewhat complicated, so I figured I'd ask if someone already did some work on this. Thanks for any pointers :)

**Comment by Myriad-Dreamin** [comment]
You can pass arguments by `sys.inputs`. I'm using this function to change inputs dynamically.

```rs
#[comemo::memoize]
fn create_library(inputs: Arc<Prehashed<Dict>>) -> Arc<Prehashed<Library>> {
    let lib = typst::Library::builder()
        .with_inputs(inputs.deref().deref().clone())
        .build();

    Arc::new(Prehashed::new(lib))
}
```

You can see a full implementation of world here. https://github.com/Myriad-Dreamin/typst.ts/blob/46cfbbe0f087510ffcba3c6e70ac4a0d2dbbd39b/compiler/src/world.rs#L106-L113

**Reply by emilyyyylime** [reply]
to expand on this, the canonical way to template Typst documents is to read the variable data from within Typst code, rather than inject it directly

**Reply by KillTheMule** [reply]
Right, that's what I'm doing, but what if, say, a user inputs a `"`, then that would read like something like

```typst
#let text = "blah " blah2"
```
which isn't valid typst, so I'd need to escape the `"` beforehand. Or am I misunderstanding you?

**Reply by emilyyyylime** [reply]
How are you doing this? If you get the text in as a native Typst string it can be embedded as `#content`, or if you do want to evaluate it with `#eval(markup, mode: "markup")`

**Reply by KillTheMule** [reply]
No it does not need to be evaluated. But it is read from a spreadsheet and needs to be inserted in typst into the right places, and it's user provided, so I basically have to assume they might type in anything :) Not sure how to make that into a "native Typst string".

**Reply by emilyyyylime** [reply]
Then there's no need to escape Typst strings, they aren't like macros or anything. If you're reading from a spreadsheet I reckon the `csv` function is what best fits your usecase

**Reply by KillTheMule** [reply]
Reading from a spreadsheet is done in Rust, I need to mangle some things quite a bit (and there's a GUI involved, too).

But I'm totally unclear on what you mean by `Typst strings` here. I feel totally dense, but can you elaborate somehow? Say you just have a trivial template document

```typst
= <Placeholder>
```

where you want the user to provide a heading. How would you go about inserting that with "Typst strings"? Maybe you could point to an example or documentation, so I can stop bugging you :) Thanks in advance for your help!

**Reply by emilyyyylime** [reply, marked as answer]
What I would do is indeed to supply arguments through `World` somehow, either as @Myriad-Dreamin mentioned with arguments which can then be read from within Typst as `#sys.args.at(...)` or with files which can be read from Typst using `#read()` etc.

This is in contrast to patching a source file with some portions hard coded and others taken from runtime values (a la `format!()`), which would require some escape-sequence considerations

**Reply by KillTheMule** [reply]
Ahh ok, that sounds clearer. So I'd have a document

```typst
= #sys.args.at('heading')
```

and then create the library in code with a `Dict` that contains the corresponding entry? I'll think about that, thanks a bunch!

**Reply by KillTheMule** [reply]
Ok, tried it, works like you say, very nice. Although after some tries I have to say that strings in the file source work very well either, they seem very error-resistant, so that the only thing I'd have to escape are quotation marks really, so I might still go that route (since I've already set up all the templating, it's not clear if it's worth that much work to change this). Anyways, thanks again!

**Reply by emilyyyylime** [reply]
I don't understand why you'd need to escape quotation marks either? You're never interpreting them are you?

**Reply by KillTheMule** [reply]
Huh, well, I'm doing templating. A very simple document template might look like

```tupst
Well hello "{{name}}"
```

and if the template engine sticks in `B"C` because that's what the user gave it, it's not going to be a valid document.

Of course, in general things are more complicated, I have several documents sharing a helper file that defines functions for document elements that need to get dictionaries passed to them that have strings as values... that might be besides the point though :)

**Reply by emilyyyylime** [reply]
I'll trust you have good reason to do what you're doing, but if you want templating why create a custom syntax rather than using native Typst? you could generate some module file that can be imported in the template and then it could look like `Well hello #name` without any pre-processing

**Reply by KillTheMule** [reply]
> I'll trust you have good reason to do what you're doing, but if you want templating why create a custom syntax rather than using native Typst?

Well, I don't :D Historically, I've been templating LaTex files (boy, that's a can of worms...), and I switched to typst just recently, so I kept on doing the thing I did before. I did not know about `inputs`, so that path wasn't really open to me.

> you could generate some module file that can be imported in the template and then it could look like Well hello #name without any pre-processing

This is where I'm still not clear. How would that module file look? I imagine something like

```typst
#let name = <namehere>
```

where I have to replace `<namehere>` by the user input. But that would leave me back at having to quote things, right?

**Reply by emilyyyylime** [reply]
Well, no, you'll replace <namehere> by `sys.inputs.at("name")` and programatically pass the inputs in. The Typst parser should never need handle the raw data

**Reply by emilyyyylime** [reply]
Or you'd use `json()`/`toml`/`csv` through an intermediary file in which case you'll need to escape inputs for the chosen format

**Reply by KillTheMule** [reply]
Yeah, ok, the `sys.inputs` route  works, I'll just have to decide if I'll make that rather largeish change in my code :) Thanks! I'd rather not introduce another intermediary format I think.

**Comment by KillTheMule** [comment]
> You can pass arguments by `sys.inputs`. I'm using this function to change inputs dynamically.
> 
> ```rust
> #[comemo::memoize]
> fn create_library(inputs: Arc<Prehashed<Dict>>) -> Arc<Prehashed<Library>> {
>     let lib = typst::Library::builder()
>         .with_inputs(inputs.deref().deref().clone())
>         .build();
> 
>     Arc::new(Prehashed::new(lib))
> }
> ```
> 
> You can see a full implementation of world here. https://github.com/Myriad-Dreamin/typst.ts/blob/46cfbbe0f087510ffcba3c6e70ac4a0d2dbbd39b/compiler/src/world.rs#L106-L113

I don't fully understand, and I did not find any documentation on that. I mean, I could certainly stick my (rather largeish) dynamic data into a `Dict` and then create this library... but how would I use this? And why would that solve the quoting problem? Sorry if I'm being dense ;)

**Comment by santialbo** [comment]
I'm not sure if it's the same usecase but this way you can "escape" user input
```typst
#{let t = "this is the text";[#t]}
```
https://typst.app/project/rmh9rSRzI2Cwwyn1-P9hg-
